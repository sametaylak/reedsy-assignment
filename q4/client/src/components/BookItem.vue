<template>
  <div class="book-item" v-if="book">
    <img class="book-cover" :src="book.cover" @click="goToDetail"/>
    <div class="book-item-info">
      <h2 class="title" @click="goToDetail">{{ order + 1}}. {{ book.title }}</h2>
      <span class="author">{{ book.author }}</span>
      <p class="synopsis">{{ book.synopsis | truncate(200, '...') }}</p>
      <UpvoteArea :book="book"/>
    </div>
  </div>
</template>

<script>
import UpvoteArea from '@/components/UpvoteArea'

export default {
  name: 'BookItem',
  components: {
    UpvoteArea
  },
  props: {
    order: Number,
    book: Object
  },
  methods: {
    goToDetail () {
      this.$router.push({
        name: 'BookDetail',
        params: { slug: this.book.slug }
      })
    }
  }
}
</script>

<style lang="scss" scoped>
.book-item {
  display: flex;
  &-info {
    margin-left: 16px;
    .title {
      cursor: pointer;
      margin: 0;
    }
    .author {
      font-style: italic;
    }
    .upvote-area {
      float: right;
    }
  }
  img {
    cursor: pointer;
    border-radius: 6px;
    width: 10%;
  }
  @media only screen and (max-width: 768px) {
    flex-direction: column;
    img {
      display: block;
      margin: 0 auto;
      width: 50%;
    }
  }
}
</style>
