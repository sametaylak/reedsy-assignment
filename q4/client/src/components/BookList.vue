<template>
  <div class="book-list">
    <BookItem v-for="(book, index) in paginatedData" :order="index + (page * perPage)" :book="book" :key="book.slug"/>
    <Pagination
      :prevPageDisableCondition="onFirstPage"
      :nextPageDisableCondition="onLastPage"
      :prevPage="prevPage"
      :nextPage="nextPage" />
  </div>
</template>

<script>
import BookItem from '@/components/BookItem'
import Pagination from '@/components/Pagination'
import paginate from '@/mixins/paginate'

export default {
  name: 'BookList',
  mixins: [
    paginate({ entity: 'list', perPage: 3 })
  ],
  components: {
    BookItem,
    Pagination
  },
  props: {
    list: {
      type: Array,
      default: () => []
    }
  }
}
</script>

<style lang="scss" scoped>
.book-list {
  .book-item:not(:last-child) {
    margin-bottom: 16px;
  }
}
</style>
