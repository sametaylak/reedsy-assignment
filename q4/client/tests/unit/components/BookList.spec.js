import { shallowMount } from '@vue/test-utils'
import BookList from '@/components/BookList.vue'
import BookItem from '@/components/BookItem.vue'

describe('BookList.vue', () => {
  it('has a list prop', () => {
    const wrapper = shallowMount(BookList)

    expect(wrapper.props().list).toEqual([]);
  })

  it('renders BookItem when props.list passed', () => {
    const list = [
      {
        author: 'Samet A.',
        title: 'Odyssey',
        slug: 'odyssey',
        synopsis: 'best book ever',
        upvoted: true,
        upvotes: 101
      }
    ]

    const wrapper = shallowMount(BookList, {
      propsData: { list }
    })

    expect(wrapper.props().list).toEqual(list)
    expect(wrapper.find(BookItem).exists()).toBe(true)
    expect(wrapper.findAll(BookItem).length).toEqual(1)
  })

  it('should have consistent dom', () => {
    const list = [
      {
        author: 'Samet A.',
        title: 'Odyssey',
        cover: 'odyssey.jpg',
        slug: 'odyssey',
        synopsis: 'best book ever',
        upvoted: true,
        upvotes: 101
      }
    ]

    const wrapper = shallowMount(BookList, {
      propsData: { list }
    })

    expect(wrapper).toMatchSnapshot()
  })
})
