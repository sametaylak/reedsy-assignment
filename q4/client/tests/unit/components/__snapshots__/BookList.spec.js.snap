// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList.vue should have consistent dom 1`] = `
<div class="book-list">
  <bookitem-stub order="0" book="[object Object]"></bookitem-stub>
</div>
`;
